### DISCUSSION TITLE
UI testing: simulator or physical device?

### DISCUSSION THREAD

**Discussion by ileitch** [MobileNativeFoundation/discussions#153] [General]
Hey all. I'm interested your experiences running large UI test suites, and specifically how you used the simulator vs physical devices?

At Thumbtack we used the simulator for PR blocking UI tests, and real devices for a small subset of pre-release smoke tests. For the PR tests, the ease of use of the simulator and the ability to cost-effectively parallelize tests was a huge benefit.

At Reddit we're just now fleshing out our UI testing strategy, so it's a good time to learn from the community. Personally, I'm skeptical that PR blocking tests can be effectively scaled and managed using physical devices, but would be very interested to hear if anyone has achieved this.

**Comment by keith** [comment]
We're using simulators only and in general I've been very wary of trying to maintain physical devices for this, definitely interested to hear if many folks are doing that successfully with a large number of devices.

**Comment by bootstraponline** [comment]
[Flank](https://github.com/Flank/flank) has run 160M tests in the last six months. 25% on real devices, 75% on virtual. Scaling pre-merge UI testing is best done with virtual devices. Firebase Test Lab is popular for scaling Android tests. For iOS, a common approach is using iOS simulators on mac minis with [bluepill](https://github.com/MobileNativeFoundation/bluepill).


- https://medium.com/pinterest-engineering/pre-submit-ui-tests-at-pinterest-556be1611be
- https://dropbox.tech/mobile/revamping-the-android-testing-pipeline-at-dropbox
